src/core/inductive.ml

```diff
@@ -175,31 +247,41 @@ let gen_rec_type : Sig_state.t -> popt -> sym -> sym list -> term =
     let f_not_rec : tbox -> tvar -> tbox = fun a _ -> a in
     let init : tbox = _Type in
     fold_cons_typ pos codom inj_var build_rec_hyp domrec dom ind_sym
-      cons_sym f_rec f_not_rec init x_str
+      cons_sym f_rec f_not_rec init x_str assoc_predicates
   in
-
-  (* STEP 3: Create the conclusion of the inductive principle *)
-  let codom ts =
-    let x = Bindlib.new_var mkfree x_str in
-    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
-    _Prod (fapp ind_sym ts) t
+  (* Very close to the function "fold_right2", but "f" is the function
+     "fold_right" and it needs a neutral element. *)
+  let rec clauses ind_list e : tbox = match ind_list with
+    | [] -> e
+    | (i,cl)::q ->
+        let res = clauses q e in
+        List.fold_right (fun a b -> _Impl (case_of i a) b) cl res
   in
-  let conclusion = gen_ind_typ_codom pos ind_sym codom x_str in
+  let t = List.map (clauses ind_list) conclusion_list in
 
-  (* STEP 4: Create the induction principle *)
-  let t =
-    List.fold_right (fun a b -> _Impl (case_of a) b) cons_list conclusion
+  (* STEP 3 - Create the induction principle. *)
+  let f t =
+    List.fold_left
+      (fun e (_,(name,typ)) -> _Prod typ (Bindlib.bind_var name e)) t
+      assoc_predicates
   in
-  let t = _Prod p_type (Bindlib.bind_var p t) in
+  let t = List.map f t in
```

Comment:
Idem here. Could you please find better names for f and t, and explain *what* `List.map f t` does. In addition, why not including Bindlib.unbox in this List.map?

Reply:
DOne